apps/champions/lib/champions/units.ex

```diff
@@ -207,6 +217,151 @@ defmodule Champions.Units do
+    {same_faction_amount, same_faction_rank} = same_faction_requirements(unit)
+
+    try do
+      validate_units(
+        unit,
+        unit_list,
+        same_character_amount,
+        same_character_rank,
+        same_faction_amount,
+        same_faction_rank
+      )
+    rescue
+      _e in RuntimeError -> false
+    end
+  end
+
+  defp same_character_requirements(%Unit{rank: @star4}), do: {2, @star4}
+  defp same_character_requirements(%Unit{rank: @star5}), do: {1, @star5}
+  defp same_character_requirements(%Unit{rank: @illumination1}), do: {1, @star5}
+  defp same_character_requirements(%Unit{rank: @illumination2}), do: {1, @star5}
+  defp same_character_requirements(%Unit{rank: @illumination3}), do: {3, @star5}
+
+  defp same_faction_requirements(%Unit{rank: @star4}), do: {4, @star4}
+  defp same_faction_requirements(%Unit{rank: @star5}), do: {4, @star5}
+  defp same_faction_requirements(%Unit{rank: @illumination1}), do: {1, @illumination1}
+  defp same_faction_requirements(%Unit{rank: @illumination2}), do: {2, @illumination2}
+  defp same_faction_requirements(%Unit{rank: @illumination3}), do: {2, @illumination2}
+
+  defp validate_units(
+         unit,
+         unit_list,
+         same_character_amount,
+         same_character_rank,
+         same_faction_amount,
+         same_faction_rank
+       ) do
+    # Remove necessary units with the same character
+    removed_same_character =
+      Enum.reduce(1..same_character_amount, unit_list, fn _, list ->
+        same_character =
+          Enum.find(
+            list,
+            &(&1.character_id == unit.character_id and &1.rank == same_character_rank)
+          )
+
+        if is_nil(same_character), do: raise("not enough of same character")
+        if same_character.user_id != unit.user_id, do: raise("consumed unit not owned")
+
+        List.delete(list, same_character)
+      end)
+
+    # Remove necessary units with the same faction
+    removed_same_faction =
+      Enum.reduce(1..same_faction_amount, removed_same_character, fn _, list ->
+        same_faction =
+          Enum.find(
+            list,
+            &(&1.character.faction == unit.character.faction and &1.rank == same_faction_rank)
+          )
+
+        if is_nil(same_faction), do: raise("not enough of same faction")
+        if same_faction.user_id != unit.user_id, do: raise("consumed unit not owned")
+
+        List.delete(list, same_faction)
+      end)
+
+    # If we got here with no raises and an empty list, then the units are valid
+    if Enum.empty?(removed_same_faction), do: true, else: raise("too many units given")
+  end
+
+  @doc """
+  Returns whether a unit can rank up, based on its current rank and its character's quality.
+  """
+  def can_rank_up(%Unit{rank: rank, character: %Character{quality: @epic}}), do: rank < @awakened
+
+  def can_rank_up(%Unit{rank: rank, character: %Character{quality: @rare}}),
+    do: rank < @illumination2
+
+  def can_rank_up(_unit), do: false
```

Comment:
Dont this need to be private?

Reply:
No! Later on we will need CanLevelUp, CanRankUp and CanTierUp messages for the frontend to display the appropriate buttons. If you think that we should make them private until then though, that's perfect